### PR TITLE
new python population genetics modules

### PR DESCRIPTION
**Pull request by alxsimon**
Adding a few python modules relating to population genetics analyses.

Nix beginner so feel free to double check my work and point out mistakes.

added modules:
- demes
- demesdraw
- tskit
- msprime
- newick
- pyslim

### CODE REVIEW

File: pkgs/development/python-modules/demes/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/demes/default.nix:18]:
```suggestion
    hash = "9e613b65b475dba277bca751de1f37a89fd5eb4d856ba27733a2099c8a823617";
```
Has the same effect but is nicer to read since it's obvious that its sha256.

File: pkgs/development/python-modules/demesdraw/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/demesdraw/default.nix:20]: same as above

File: pkgs/development/python-modules/msprime/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/msprime/default.nix:22]: same as above

File: pkgs/development/python-modules/tskit/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/tskit/default.nix:20]: same here

File: pkgs/development/python-modules/tskit/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/tskit/default.nix:40]:
```suggestion
    description = "The tree sequence toolkit";
```

File: pkgs/development/python-modules/pyslim/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/pyslim/default.nix:37]:
```suggestion
    description = "Tools for dealing with tree sequences coming to and from SLiM";
```

File: pkgs/development/python-modules/pyslim/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/pyslim/default.nix:19]: same here

File: pkgs/development/python-modules/newick/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/newick/default.nix:28]:
```suggestion
    description = "python package to read and write the Newick format";
```

File: pkgs/development/python-modules/newick/default.nix
Comment (Janik-Haag) [inline pkgs/development/python-modules/newick/default.nix:16]: same here

**Review by Janik-Haag — CHANGES_REQUESTED** [review]
Very nice PR :heart: 

Can you fix the few nitpicks and make the PRs follow `contributing.md`?
example for package init: `$pname: init at $version`
and adding your maintainer should be `maintainers: add alxsimon`

**Comment by alxsimon** [comment]
Argh I messed up my commit history while trying to rebase, I think I'll just redo from scratch.

**Comment by Janik-Haag** [comment]
You can use git reflog to revert the rebase

**Comment by alxsimon** [comment]
So with the hash change I now have this error locally

```
error: hash '9e613b65b475dba277bca751de1f37a89fd5eb4d856ba27733a2099c8a823617' does not include a type, nor is the type otherwise known from context
```

**Comment by alxsimon** [comment]
should be ready once the checks pass

**Comment by Janik-Haag** [comment]
> should be ready once the checks pass

almost you still have the "[correct hashes for demes, demesdraw, msprime, newick, pyslim and tskit](https://github.com/NixOS/nixpkgs/pull/240183/commits/2ae8120f1845e8f8617adc302f351ba4a6ab4886)" commit and nixpkgs trys to have a atomic commit history,  since you are initializing all those packages in this pr you should remove that commit and just put the correct hashes in the specific commit. Sorry for the trouble.

**Comment by alxsimon** [comment]
yeah figure this might be an issue, thanks for your patience. At least I learned new tricks in git!

**Comment by Janik-Haag** [comment]
> yeah figure this might be an issue, thanks for your patience.

No problem, I'm usually quite patient  as long as no one is getting rude or something :D 

> At least I learned new tricks in git!

hehe, I also gained most of my git experience from doing nixpkgs, For example: how to bisect to find a broken commit or how to rebase properly, etc ...

**Review by Janik-Haag — APPROVED** [review]
LGTM

**Comment by Janik-Haag** [comment]
Let's wait for CI to build successfully and after that give it like a week or so, so other people get a chance to review the changes as well :D

**Comment by alxsimon** [comment]
Thanks!

**Comment by natsukium** [comment]
Would you please add `pytestCheckHook`, etc. to `nativeCheckInputs`, as the tests may not run?

Then the commit message should be `python310Packages.xxx` for ofborg to build it properly.
see, https://github.com/NixOS/nixpkgs/pull/240343

File: pkgs/development/python-modules/tskit/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/tskit/default.nix:26]:
```suggestion
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
checkInputs are missing in all packages

**Comment by Janik-Haag** [comment]
There is https://nixos.org/manual/nixpkgs/stable/#python and https://nixos.wiki/wiki/Python#Contribution_guidelines

**Comment by alxsimon** [comment]
So one issue I have is that most of those packages either do not include the tests in the Pypi release or only part of it (mostly due to the size of test files).
What's the best strategy here?

**Comment by natsukium** [comment]
We can fetch the source from a VCS, such as GitHub, if the test is not included in the pypi source.

**Comment by alxsimon** [comment]
hmm, any idea why I have the following error during the tests for the msprime module? `ModuleNotFoundError: No module named 'msprime._msprime'`
Feels like the tests are not run in the right folder

**Comment by natsukium** [comment]
> Feels like the tests are not run in the right folder

Can you try the following?
1. add `pytestFlagsArray` and set pytest arguments
2. add `preCheck` and change the directory in it.

**Comment by alxsimon** [comment]
I think I stumble on something more complex than I can handle.
packages msprime and tskit happen to have C bindings, which seem to need special care https://github.com/NixOS/nixpkgs/issues/40835#issuecomment-1265306633
If you know a clear guide on how to do that this would be much appreciated.

**Comment by alxsimon** [comment]
Although the msprime module import `pythonImportCheck` works fine, so maybe I just have to disable the tests?

**Comment by natsukium** [comment]
I suspect that msprime will not work as it is, even if it can be imported.
I will check it, as it may not compile the C source code correctly.

**Comment by alxsimon** [comment]
thanks, just pushed my recent changes as a starter

File: pkgs/development/python-modules/msprime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/msprime/default.nix:41]:
```suggestion
    pytest
```
Exceptionally, we need to call `pytest` directly here, not `python -m pytest`. 
[Implementation](https://github.com/NixOS/nixpkgs/blob/37e60e78c8dcbc9d047dd7bed04b45b70aa20387/pkgs/development/interpreters/python/hooks/pytest-check-hook.sh#L53)

File: pkgs/development/python-modules/msprime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/msprime/default.nix:44]:
```suggestion
  checkPhase = ''
    runHook preCheck
    
    # avoid adding the current directory to sys.path
    # https://docs.pytest.org/en/7.1.x/explanation/pythonpath.html#invoking-pytest-versus-python-m-pytest
    pytest -xvs --import-mode append \
      --ignore=tests/test_demography.py \
      ...
      
    runHook postCheck
  '';
```

File: pkgs/development/python-modules/msprime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/msprime/default.nix:48]:
```suggestion
```

**Review by natsukium — COMMENTED** [review]
The following two errors have been fixed
1. gsl was not linked at compile time.
2. pytest was adding extra paths to the search path at runtime.

Regarding 2, there is a [specification](https://docs.pytest.org/en/7.1.x/explanation/pythonpath.html#invoking-pytest-versus-python-m-pytest) that the current directory is added to the head of sys.path when using pytestCheckHook, which internally calls `python -m pytest`.
In most cases, this is not a problem.
In this case, however, it was a problem.
The compiled shared object (_msprime) exists in `$out/${python.sitePackages}`, but not in the current directory used by checkPhase.
I think the above error occurs because the module in the current directory where _msprime does not exist is called as msprime.

File: pkgs/development/python-modules/msprime/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/msprime/default.nix:57]:
```suggestion
    license = licenses.gpl3Plus;
```
https://github.com/tskit-dev/msprime/blob/52abee1ed132ef18c0ad27c3b61be67e3f399a48/msprime/__init__.py#L11-L12

**Comment by alxsimon** [comment]
Thank you so much for your help @natsukium, I'll try to finish up the last package and squash some commits together

**Comment by alxsimon** [comment]
I deactivated the tests for tskit, I think this will do for now.
The reason is that tests are not included in the Pypi but the source install from GitHub is too complex for me to handle right now.
I can revise that either when I convince the upstream to include tests in Pypi or when I manage to understand more complex install.

File: pkgs/development/python-modules/newick/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/newick/default.nix:19]:
```suggestion
    rev = "v${version}";
```

File: pkgs/development/python-modules/newick/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/newick/default.nix:20]:
```suggestion
    hash = "TxyR6RYvy2oIcDNZnHrExtPYGspyWOtZqNy488OmWwk=";
```

File: pkgs/development/python-modules/newick/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/newick/default.nix:37]:
```suggestion
    description = "A python package to read and write the Newick format";
```
The first charactor of the description must be capitalized.
https://nixos.org/manual/nixpkgs/stable/#var-meta-description

File: pkgs/development/python-modules/pyslim/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/pyslim/default.nix:42]:
```suggestion
    license = licenses.mit;
```
https://github.com/tskit-dev/pyslim/blob/main/LICENSE

**Review by natsukium — COMMENTED** [review]
Thanks for accepting the checkInputs changes.
I think it's good for now.
Could you please check a few additional comments?

**Comment by alxsimon** [comment]
ok I think I resolved everything and finished squashing some commits.
Thanks again for the help.

**Review by natsukium — APPROVED** [review]
LGTM.
Thanks for your contributions. I'm glad to help to add these biological libraries in nixpkgs.

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 240183` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.msprime</li>
    <li>python311Packages.msprime.dist</li>
    <li>python311Packages.pyslim</li>
    <li>python311Packages.pyslim.dist</li>
  </ul>
</details>
<details>
  <summary>20 packages built:</summary>
  <ul>
    <li>python310Packages.demes</li>
    <li>python310Packages.demes.dist</li>
    <li>python310Packages.demesdraw</li>
    <li>python310Packages.demesdraw.dist</li>
    <li>python310Packages.msprime</li>
    <li>python310Packages.msprime.dist</li>
    <li>python310Packages.newick</li>
    <li>python310Packages.newick.dist</li>
    <li>python310Packages.pyslim</li>
    <li>python310Packages.pyslim.dist</li>
    <li>python310Packages.tskit</li>
    <li>python310Packages.tskit.dist</li>
    <li>python311Packages.demes</li>
    <li>python311Packages.demes.dist</li>
    <li>python311Packages.demesdraw</li>
    <li>python311Packages.demesdraw.dist</li>
    <li>python311Packages.newick</li>
    <li>python311Packages.newick.dist</li>
    <li>python311Packages.tskit</li>
    <li>python311Packages.tskit.dist</li>
  </ul>
</details>

**Comment by Janik-Haag** [comment]
A few packages fail to build :/

File: pkgs/development/python-modules/pyslim/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/pyslim/default.nix:33]:
```suggestion
  # Requires non-packaged software SLiM
  doCheck = false;
```

File: pkgs/development/python-modules/newick/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/newick/default.nix:29]:
```suggestion
```
we don't want to collect those, please patch out flags that make this a requirement

File: pkgs/development/python-modules/msprime/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/msprime/default.nix:54]: Please use the pytestCheckHook

File: pkgs/development/python-modules/demesdraw/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/demesdraw/default.nix:40]:
```suggestion
```

File: pkgs/development/python-modules/demes/default.nix
Comment (SuperSandro2000) [inline pkgs/development/python-modules/demes/default.nix:36]:
```suggestion
```

File: pkgs/development/python-modules/msprime/default.nix
Comment (alxsimon) [reply pkgs/development/python-modules/msprime/default.nix:54]: this is related to @natsukium's change https://github.com/NixOS/nixpkgs/pull/240183#discussion_r1247847237
From what I understood this is a workaround so that tests can be run in this specific case.

**Comment by alxsimon** [comment]
> A few packages fail to build :/

this should be resolved as I removed two tests in msprime that were failing with 3.11.
pyslim error was linked to msprime dependency.

File: pkgs/development/python-modules/msprime/default.nix
Comment (alxsimon) [reply pkgs/development/python-modules/msprime/default.nix:60]: So I've just tried using the flag `--import-mode append` inside `pytestFlagsArray` and that doesn't work with `pytestCheckHook`

Comment (natsukium) [reply pkgs/development/python-modules/msprime/default.nix:60]: It seems that this is a fairly rare case where pytestCheckHook does not work due to Python's complex import system.

**Comment by Janik-Haag** [comment]
Result of `nixpkgs-review pr 240183` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>24 packages built:</summary>
  <ul>
    <li>python310Packages.demes</li>
    <li>python310Packages.demes.dist</li>
    <li>python310Packages.demesdraw</li>
    <li>python310Packages.demesdraw.dist</li>
    <li>python310Packages.msprime</li>
    <li>python310Packages.msprime.dist</li>
    <li>python310Packages.newick</li>
    <li>python310Packages.newick.dist</li>
    <li>python310Packages.pyslim</li>
    <li>python310Packages.pyslim.dist</li>
    <li>python310Packages.tskit</li>
    <li>python310Packages.tskit.dist</li>
    <li>python311Packages.demes</li>
    <li>python311Packages.demes.dist</li>
    <li>python311Packages.demesdraw</li>
    <li>python311Packages.demesdraw.dist</li>
    <li>python311Packages.msprime</li>
    <li>python311Packages.msprime.dist</li>
    <li>python311Packages.newick</li>
    <li>python311Packages.newick.dist</li>
    <li>python311Packages.pyslim</li>
    <li>python311Packages.pyslim.dist</li>
    <li>python311Packages.tskit</li>
    <li>python311Packages.tskit.dist</li>
  </ul>
</details>

**Comment by Janik-Haag** [comment]
I think we can merge this once ofborg was successful, @natsukium what do you think?

File: pkgs/development/python-modules/demesdraw/default.nix
Comment (natsukium) [inline pkgs/development/python-modules/demesdraw/default.nix:43]:
```suggestion
    # This variable is needed to suppress the "Trace/BPT trap: 5" error in Darwin's checkPhase.
  # Not sure of the details, but we can avoid it by changing the matplotlib backend during testing.
  env.MPLBACKEND = lib.optionalString stdenv.isDarwin "Agg";
```

**Review by natsukium — COMMENTED** [review]
I think it is fine.
Finally, could you please comment on the section I have added?
Because it is unclear what the variable is needed for and what it does.
I apologize. I forgot to comment on this.

**Comment by Janik-Haag** [comment]
Thank you a lot @alxsimon for the PR and @natsukium for the through review. 

Off-topic: natsukium your email seems to be broken, it at least bounced mine saying it doesn't exist :shrug: 
Do you have a more privet/direct way to contact you like matrix?

File: pkgs/development/python-modules/msprime/default.nix
Comment (SuperSandro2000) [reply pkgs/development/python-modules/msprime/default.nix:60]: pytestCheckHook is just some shell scripting around pytest. It should really work with this, this is more about making the ignores below easier to manage.

**Pull request merged**
